File: Sorting Algorithms/Heap Sort.cpp
#include <iostream>

using namespace std;

void heapSort( int A[], int N );
void pushHeap( int heap[], int data, int N );
int popHeap( int heap[], int N );

int main()
{
    int N;

    cout << "Enter No of Elements in array : ";
    cin >> N;

    int arr[N];

    cout << "Enter Elements of array : ";
    for ( int  i = 0; i < N; i++ )
    {
        cin >> arr[i];
    }

    heapSort( arr, N );

    cout << "Elements in Sorted order are " << endl;
    for ( int i = 0; i < N; ++i )
    {
        cout << arr[i] << " ";
    }

    return 0;
}

void heapSort( int A[], int N )
{
    int heap[N];
    int heapEnd =  -1;

    for( heapEnd; heapEnd < N; heapEnd++ )
    {
        pushHeap( heap, A[heapEnd], heapEnd );
    }


    while( heapEnd >= 0 )
    {
        A[heapEnd] = popHeap ( heap, heapEnd );
        heapEnd--;
    }
}

void pushHeap( int heap[], int data, int N )
{
    N++;
    heap[N] = data;

    while( N > 0 )
    {
        int parent = ( N - 1 ) / 2;
        if( heap[parent] < heap[N] )
        {
            int temp = heap[parent];
            heap[parent] = heap[N];
            heap[N] = temp;
            N = parent;
        }else
        {
            break;
        }
    }

}

int popHeap( int heap[], int N )
{
    int max = heap[0];

    heap[0] = heap[N];
    N--;

    int i  = 0, leftchild, rightchild;

    while( i < N )
    {
        leftchild = ( 2 * i ) + 1;
        rightchild = ( 2 * i ) + 2;

        if( heap[leftchild] > heap[i] && heap[leftchild] > heap[rightchild] && leftchild <= N )
        {
            int temp = heap[leftchild];
            heap[leftchild] = heap[i];
            heap[i] = temp;
            i = leftchild;
        }else if ( heap[rightchild] > heap[i] && rightchild <= N )
        {
            int temp = heap[rightchild];
            heap[rightchild] = heap[i];
            heap[i] = temp;
            i = rightchild;
        }else
        {
            break;
        }
    }

    return max;
}

File: Sorting Algorithms/Python/Heap Sort.py
def heapSort( arr, len ):
    heap = []
    heapend = -1
    for i in range( len ):
        pushHeap( heap, arr[i], heapend )
        heapend += 1

    while heapend >= 0:
        arr[heapend] = popHeap(heap, heapend)
        heapend -= 1

def pushHeap( heap, data, heapend):
    heapend += 1
    heap.append( data )

    while heapend > 0 :
        parent = int( ( heapend - 1 ) / 2 )
        if heap[parent] < heap[heapend] :
            heap[parent], heap[heapend] = heap[heapend], heap[parent]
            heapend = parent
        else :
            break

def popHeap( heap, heapend ) :
    max = heap[0]
    heap[0] = heap[heapend]
    heap.pop()
    heapend -= 1

    i = 0

    while i < heapend :
        leftchild = int( ( 2 * i ) + 1 )
        rightchild = int( ( 2 * i ) + 2 )

        if leftchild <= heapend and  rightchild <= heapend:
            if heap[leftchild] > heap[i] and heap[leftchild] > heap[rightchild]:
                heap[leftchild], heap[i] = heap[i], heap[leftchild]
                i = leftchild
            elif heap[rightchild] > heap[i]:
                heap[rightchild], heap[i] = heap[i] ,heap[rightchild]
                i = rightchild
            else :
                break
        elif leftchild <= heapend and rightchild > heapend :
            if heap[leftchild] > heap[i]:
                heap[leftchild], heap[i] = heap[i], heap[leftchild]
                i = leftchild
            else:
                break
        else:
            break
    return max


len = int( input( "Enter length of array : " ) )

arr = []

for i in range( len ):
    arr.append(int( input( "Enter value in array : " ) ) )

heapSort( arr, len )

print( "Sorted array is :", arr )
